maven-core/src/main/java/org/apache/maven/plugin/internal/ValidatingConfigurationListener.java

```diff
@@ -80,18 +95,83 @@ public void notifyFieldChangeUsingReflection( String fieldName, Object value, Ob
     {
         delegate.notifyFieldChangeUsingReflection( fieldName, value, target );
 
-        if ( mojo == target )
+        if ( mojo == target && value != null )
         {
             notify( fieldName, value );
         }
     }
 
     private void notify( String fieldName, Object value )
     {
-        if ( value != null )
+        missingParameters.remove( fieldName );
+
+        if ( logger.isWarnEnabled() )
+        {
+            warnDeprecated( fieldName, value );
+        }
+    }
+
+    private void warnDeprecated( String fieldName, Object value )
+    {
+        Parameter parameter = mojoDescriptor.getParameterMap().get( fieldName );
+        String deprecated = parameter.getDeprecated();
+        if ( deprecated != null && !deprecated.isEmpty() )
+        {
+            Object defaultValue = evaluateValue( parameter.getDefaultValue() );
+            if ( !toString( value ).equals( toString( defaultValue ) ) )
+            {
+                StringBuilder sb = new StringBuilder( "  Parameter '" );
+                sb.append( fieldName ).append( '\'' );
+                if ( parameter.getExpression() != null )
+                {
+                    String userProperty = parameter.getExpression().replace( "${", "'" ).replace( '}', '\'' );
+                    sb.append( " (User Property " ).append( userProperty ).append( ")" );
+                }
+                sb.append( " is deprecated. " ).append( deprecated );
```

Comment:
You haven't even evaluated the content of `deprecated`, how do you know then?

Reply:
`deprecated` variable is taken from `parameter.getDeprecated();`
all magic here is to prepare message like:
```
[WARNING]   Parameter 'deprecatedParam2' (User Property 'config.deprecatedParam2') is deprecated.
```

Reply:
This still doesn't make sense. The API says:

    @param deprecated <code>true</code> to deprecate the Mojo, <code>false</code> otherwise.

So the output will read: ...is deprecated. true"

Why not `"true"equals(deprecated)`?

Reply:
`parameter.getDeprecated()` contains reason of depracation from plugin documentation.
message will be:
```
is deprecated. Reason provided by author ...
```
please look at IT tests

Reply:
I'd turn the period onto a colon.